Jrhosk/1837 optional spectral profiler info

jrhosk opened the pull request:
**Description**
Implemented optional cursor information for spectral profile. Optional display can be toggled via spectral profile settings.


(updates by Yu-Hsuan)
Closes #1837. Please note that currently we do not support secondary unit for the smoothed data. This needs extra calculation, and I prefer we file a new issue if needed.
Changes:
* Added `channelSecondaryValues`, `spectralTypeSecondary`, and `spectralUnitSecondary` for the converted x data of the spectral profile.
* Added toggle and dropdown in the spectral profiler settings to enable the feature.
* Implemented `toFormattedNotationByDiff` to calculate the required digits of the cursor info (see https://github.com/CARTAvis/carta-frontend/pull/1883#issuecomment-1171323434 for the details).

Minor change:
* Renamed `startEndIndexes` to `dataIndexes` in `SpectralProfileWidgetStore`

**Checklist**

For linked issues (if there are):
- [X] assignee and label added
- [X] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [X] reviewers and assignee added
- [X] ZenHub estimate, milestone, and release (if needed) added
- [X] e2e test passing / ~corresponding fix added~
- [X] changelog updated / ~no changelog update needed~
- [X] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [X] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

jrhosk committed:
Implementation of additional axis info in spectral profile cusror info.

jrhosk committed:
Further progress on reimplimenting optional cursor values.

jrhosk committed:
Optional cursor values implemented. Things left to do :: clean up code; ensure variable names are consistent; finish minor issues mention by Yu.

jrhosk committed:
implementation of optional cursor information is finiished aside from one small bug that I will consult Yuhsuan about..

jrhosk committed:
Implemented fix for bug where optional cursor values did not update properly when zooming.

jrhosk committed:
Changed requested by Yu.

jrhosk committed:
Add new inputs to cursor generation function.

jrhosk committed:
Merge branch 'dev' into jrhosk/1837-optional-spectral-profiler-info

jrhosk commented:
@YuHsuan-Hwang I made some change according to our slack discussion. I think this is ready for testing.

kswang1029 commented:
@jrhosk I had a quick look and found an issue about channel index. If we set the main axis as channel and the 2nd axis as velocity, when we zoom in a profile we see the displayed channel index for a specific channel is changing with different zoom levels. It appears that the channel indices count the channels in the field of view only. The channel index per channel should be invariant at different profile zoom levels. 


https://user-images.githubusercontent.com/20819712/176071279-e7233be7-f6b5-4b0f-a084-978a0b8f4a75.mov

jrhosk committed:
Fixed bug where CHANNEL was not stable when zooming in Spectral Profile.

jrhosk committed:
Consolidate if-else statement per Yu's requests.

jrhosk commented:
> @jrhosk I had a quick look and found an issue about channel index. If we set the main axis as channel and the 2nd axis as velocity, when we zoom in a profile we see the displayed channel index for a specific channel is changing with different zoom levels. It appears that the channel indices count the channels in the field of view only. The channel index per channel should be invariant at different profile zoom levels.
> 
>  Screen.Recording.2022-06-28.at.09.47.51.mov

Hi @kswang1029 I fixed this. I am still testing a bit on my side but I think it is about ready to switch to full testing.

jrhosk commented:
Yu asked me to include an explanation of the difference between `formattedNoation` and `toFormattedNotation`. The main difference is that the former should provide dynamic precision versus the prior which has a few levels of exponential precision.
I added the new function because I found that in some cases the cursor rounding was such that there was no change in the displayed values between ticks, i.e. the cursor shows 1.3 when the ticks are varying as 1.3x. The new function bases the precision on the delta between the nearest tick values by iterating through the decimal value and comparing each decimal place to determine the minimum precision needed. It also trims additional trailing zeros. This makes the displayed values a bit longer but beyond that seems to work in initial testing. I look forward to seeing what others think.

YuHsuan-Hwang committed:
Merge branch 'dev' into jrhosk/1837-optional-spectral-profiler-info

kswang1029 commented:
@jrhosk I have some manual tests and have run e2e test. The e2e tests show some test failures due to the change of the displayed number of digits which are understandable based on what you have described. For this, can we force to display two extra digits based on what you have derived? This is to avoid confusions of rounding error of adjacent channels. In some cases, for example, we see the step is 0.1 but some times 0.2 due to rounding differences.

Other issues spotted:
1. the displayed value of the first channel is not processed to display proper number of digits. Only the 2nd channel and so on, including the last channel are processed. 

https://user-images.githubusercontent.com/20819712/178406813-c5f1c2f7-3d56-4a0d-a80b-c19dce44a446.mov

2. when the secondary cursor info is enabled, we see an extra "channel" string to the 2nd cursor value.
![Screen Shot 2022-07-12 at 12 13 58](https://user-images.githubusercontent.com/20819712/178407170-eb90e251-ec57-4be8-9488-e0eada459cb7.png)

acdo2002 commented on src/stores/Frame/FrameStore.ts:
```diff
@@ -914,6 +936,11 @@ export class FrameStore {
         }
     }
 
+    // return spectralFrame
```
Could you explain this public returnSpectralFrame()? Because I don't understand why do we need this?

jrhosk replied:
Good catch, it is not needed anymore. I just missed it after refactoring recently. It is now removed.

acdo2002 reviewed (commented):
I have the same comment to Kuo-Song's second comment.
It quite confuse me the secondary Info shows "channel", I suggest change the label to "Secondary Info" or show the current "SpectralType (e.g. VRAD, VOPT...etc)". Either way would be more clear to me.

jrhosk commented:
> I have the same comment to Kuo-Song's second comment. It quite confuse me the secondary Info shows "channel", I suggest change the label to "Secondary Info" or show the current "SpectralType (e.g. VRAD, VOPT...etc)". Either way would be more clear to me.

This wasn't a feature, it was a bug. After consolidating the output string this got lost in the logic. It should be fixed now.

jrhosk commented:
> @jrhosk I have some manual tests and have run e2e test. The e2e tests show some test failures due to the change of the displayed number of digits which are understandable based on what you have described. For this, can we force to display two extra digits based on what you have derived? This is to avoid confusions of rounding error of adjacent channels. In some cases, for example, we see the step is 0.1 but some times 0.2 due to rounding differences.
> 
> Other issues spotted:
> 
> 1. the displayed value of the first channel is not processed to display proper number of digits. Only the 2nd channel and so on, including the last channel are processed.
> 
>  Screen.Recording.2022-07-12.at.12.06.06.mov 
> 2. when the secondary cursor info is enabled, we see an extra "channel" string to the 2nd cursor value.
>    ![Screen Shot 2022-07-12 at 12 13 58](https://user-images.githubusercontent.com/20819712/178407170-eb90e251-ec57-4be8-9488-e0eada459cb7.png)

@kswang1029 I am pushing an updated version now that I think fixes these issues. While I don't understand your e2e issues, I added two decimal places to the precision and hopefully that will fix your issue.

jrhosk committed:
Address issues with decimal precision mentioned in Kusong's review. Fix channel string issue and remove unneeded functions from FrameStore.

kswang1029 commented:
@jrhosk thanks for addressing the issues found in my previous tests. Just spotted one more:
In some cases, the wavelength in mm as the 2nd info does not have sufficient digits to distinguish adjacent channels. 

https://user-images.githubusercontent.com/20819712/178889441-7ebe3050-306e-4706-a087-9687000c81ea.mov

YuHsuan-Hwang commented on src/components/Shared/SpectralSettings/SpectralSettingsComponent.tsx:
```diff
@@ -39,6 +42,17 @@ export class SpectralSettingsComponent extends React.Component<{
                     />
                 </FormGroup>
+                {this.props.secondaryAxisCursorInfoVisible && (
+                    <FormGroup label={"Opt. Coord."} inline={true} disabled={disableCoordinateSetting}>
```
I personally prefer we use "Secondary Coordinate" for label. We can increase the label width if necessary.

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -35,6 +36,8 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
         };
     }
 
+    private profileIndex: number = 0;
```
Please remove unused variable.

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
 
-    private genCursoInfoString = (data: Point2D[], cursorXValue: number, cursorXUnit: string, label: string): string => {
-        let cursorInfoString = undefined;
+    private precisionFormatting = (nearest: {point: Point2D; index: number}, data: number, diff: number, spectralType: SpectralType): string => {
```
Looks like we can remove `nearest` in this function.

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
-    private genCursoInfoString = (data: Point2D[], cursorXValue: number, cursorXUnit: string, label: string): string => {
-        let cursorInfoString = undefined;
+    private precisionFormatting = (nearest: {point: Point2D; index: number}, data: number, diff: number, spectralType: SpectralType): string => {
+        if (spectralType === SpectralType.CHANNEL) {
```
can be simplified to `return spectralType === SpectralType.CHANNEL ?? toFixed(data) : toFormattedNotation(data, diff);`

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+            primaryXStr = this.precisionFormatting(nearest, data[nearest.index].x, diffLeft, frame.spectralType);
+
+            if (this.widgetStore.secondaryAxisCursorInfoVisible) {
+                let secondary = this.widgetStore.effectiveFrame.channelSecondaryValues.slice(start, end);
```
can be simplified to `frame`

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+
+                // Use precision to determine the proper rounding and zero suppression for displayed value. Data and secondary
+                // are handled idfferently because they have different structures.
+                diffLeft = nearest.index - 1 >= 0 ? Math.abs(data[nearest.index].x - data[nearest.index - 1].x) : Math.abs(data[nearest.index + 1].x - data[nearest.index].x);
```
shall `data` here be `secondary` instead?

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+
+        let diffLeft: number = undefined;
+        let secondaryXUnit = "";
+        let channelString = "";
```
maybe rename to `secondaryChannelString`

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,21 +155,59 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+                if (frame.spectralTypeSecondary !== SpectralType.CHANNEL) secondaryXUnit = frame.spectralUnitSecondary;
+                else channelString = "Channel ";
+
+                const xLabel =
```
Maybe we can apply `let xLabel = primary string` before if block and then `xLabel += secondary string` in the if block to avoid duplicate code.

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerSettingsPanelComponent/SpectralProfilerSettingsPanelComponent.tsx:
```diff
@@ -72,6 +77,14 @@ export class SpectralProfilerSettingsPanelComponent extends React.Component<Widg
         });
     }
 
+    @action setCursorInfo(state: boolean) {
```
remove unused functions `setCursorInfo` and `handleSecondaryAxisCursorInfoChanged`

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerSettingsPanelComponent/SpectralProfilerSettingsPanelComponent.tsx:
```diff
@@ -22,6 +22,8 @@ export enum SpectralProfilerSettingsTabs {
 
 @observer
 export class SpectralProfilerSettingsPanelComponent extends React.Component<WidgetProps> {
+    secondaryAxisCursorInfoVisible?: boolean;
```
remove unused variable

jrhosk replied:
This isn't unused. It is used in 6 places in that file.

YuHsuan-Hwang replied:
Those are either for initialization, or in unused functions, or `widgetStore.secondaryAxisCursorInfoVisible`.

YuHsuan-Hwang commented on src/stores/Frame/FrameStore.ts:
```diff
@@ -698,6 +711,15 @@ export class FrameStore {
         return label;
     }
 
+    @computed get secondarySpectralLabel(): string {
```
remove unused variable

YuHsuan-Hwang commented on src/stores/Frame/FrameStore.ts:
```diff
@@ -110,6 +110,7 @@ export class FrameStore {
     public readonly colorbarStore: ColorbarStore;
 
     public spectralCoordsSupported: Map<string, {type: SpectralType; unit: SpectralUnit}>;
+    public secondarySpectralCoordsSupported: Map<string, {type: SpectralType; unit: SpectralUnit}>;
```
perhaps we can remove `secondarySpectralCoordsSupported` and use `spectralCoordsSupported` for both

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,88 @@ export function toFixed(val: number, decimals: number = 0): string {
 }
 
+export function trimTrailingDecimals(value: string): string {
+    var splitValue: string[] = value.split(".");
```
Please use `let` instead of `var`.

YuHsuan-Hwang reviewed (changes requested):
I listed some suggestions. Most of them should be minor.

jrhosk committed:
Requested minor formatting changes from Yu.

jrhosk commented:
I think I have fixed all the issues listed and added Yu's format changes. A quick tests shows things working.

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,88 @@ export function toFixed(val: number, decimals: number = 0): string {
+    // up to 14 decimal places to determine the approxmiate
+    // precision.
+
+    var decimalPlacement = 0.1;
```
Shall `decimalPlacement` be 0.001 because `precision` is 3?
Please use `let/const` instead of `var` for all the variables in this file.

jrhosk replied:
> Shall `decimalPlacement` be 0.001 because `precision` is 3? Please use `let/const` instead of `var` for all the variables in this file.

Yes, I think I changed this after trying to fix another error that was quoted that came about after some of the cosmetic changes you requested. I'll fix it.

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,88 @@ export function toFixed(val: number, decimals: number = 0): string {
+    var totalPrecision = frontDecimalLength + precision;
+
+    // Trim trailing zeros
+    var trimmedValue = trimTrailingZeros(value.toPrecision(totalPrecision));
```
Maybe we can use `toFix` instead of `toPrecision`, so we don't need to calculate `frontDecimalLength`.

jrhosk replied:
> Maybe we can use `toFix` instead of `toPrecision`, so we don't need to calculate `frontDecimalLength`.

I will need to look at it again but there was a reason I did this. If I recall, using toFixed() screws up the channel values formatting.

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,88 @@ export function toFixed(val: number, decimals: number = 0): string {
+    // Trim trailing zeros
+    var trimmedValue = trimTrailingZeros(value.toPrecision(totalPrecision));
+
+    return value < 1 ? trimTrailingDecimals(trimmedValue) : trimTrailingDecimals(trimmedValue);
```
Can be simplify to `return trimTrailingDecimals(trimmedValue);`.
Perhaps we don't need to apply `trimTrailingDecimals` here because we have removed the zeros in `trimTrailingZeros`?

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerSettingsPanelComponent/SpectralProfilerSettingsPanelComponent.tsx:
```diff
@@ -1,7 +1,7 @@
 import * as React from "react";
-import {computed, autorun} from "mobx";
+import {computed, autorun, /*observable,*/ makeObservable} from "mobx";
```
remove comment code.

YuHsuan-Hwang commented on src/components/SpectralProfiler/SpectralProfilerSettingsPanelComponent/SpectralProfilerSettingsPanelComponent.tsx:
```diff
@@ -58,6 +58,8 @@ export class SpectralProfilerSettingsPanelComponent extends React.Component<Widg
     constructor(props: WidgetProps) {
         super(props);
+        makeObservable(this);
+
```
Do we also remove `makeObservable`?

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,88 @@ export function toFixed(val: number, decimals: number = 0): string {
+    return precision;
+}
+
+export function toFormattedNotation(value: number, delta: number): string {
```
We need a better naming to distinguish `toFormattedNotation` and `formattedNotation`. Perhaps `formattedNotationByDiff`?

jrhosk replied:
I would like to suggest a best practices document for these things. If there was a documented naming convention it would have saved half the changes requests. The purpose of naming it `toFormattedNotation` was to match the standard `toFixed` format, which I think is actually better than `formattedNotation`. I had intended to remove the later if my function was ok'd.

I'll make all these changes on Monday when I am back in the office.

YuHsuan-Hwang reviewed (changes requested):
The main changes look good to me 👍
I just have some questions related to formatting. Seems that after changing from  `toFormattedNotation` to `formattedNotation`, we don't use scientific notation. Not sure if this affects user experience.

jrhosk committed:
Add changes requests by Yu.

jrhosk committed:
Push changes from unit.ts. Missed it on the last push.

kswang1029 commented:
We have a set of e2e tests for the cursor info check. There are some test failures due to less significant digits.
The test would expect
`Cursor: (86.748221 GHz, 1.69e-2)
`but it sees
`Cursor: (86.7482 GHz, 1.69e-2)
`
The channel width from the image header is
`CDELT3 = -2.442377011414E+05 Hz
`
Or the test would expect
`Cursor: (-2.972 km/s, -5.99e-4)
`but it sees
`Cursor: (-3 km/s, -5.99e-4)
`
The channel width from the image header is
`CDELT3 = -2.441261094971E+05 Hz
`
The rest frequency is
`RESTFRQ = 2.203986800000E+11 Hz
`

I wonder if we can add additional digits (3) on top of the current approach? The reason is there are two groups of radio astronomers who are used to MHz or GHz, respectively. When it is MHz, usually we display the value with 3 or 4 decimal digits. So when we display in GHz, we need more digits (the above example). We cannot use fixed number of decimal digits as in some cases, the channel width can be very small so it is possible that adjacent channels would result in identical displayed value. So the solution I am proposing is to add additional 3 digits on top of the current approach.

Note there is also one e2e test failure about region matching. Merging dev to this branch should make the test passing.

jrhosk commented:
> We have a set of e2e tests for the cursor info check. There are some test failures due to less significant digits. The test would expect `Cursor: (86.748221 GHz, 1.69e-2) `but it sees `Cursor: (86.7482 GHz, 1.69e-2) ` The channel width from the image header is `CDELT3 = -2.442377011414E+05 Hz ` Or the test would expect `Cursor: (-2.972 km/s, -5.99e-4) `but it sees `Cursor: (-3 km/s, -5.99e-4) ` The channel width from the image header is `CDELT3 = -2.441261094971E+05 Hz ` The rest frequency is `RESTFRQ = 2.203986800000E+11 Hz `
> 
> I wonder if we can add additional digits (3) on top of the current approach? The reason is there are two groups of radio astronomers who are used to MHz or GHz, respectively. When it is MHz, usually we display the value with 3 or 4 decimal digits. So when we display in GHz, we need more digits (the above example). We cannot use fixed number of decimal digits as in some cases, the channel width can be very small so it is possible that adjacent channels would result in identical displayed value. So the solution I am proposing is to add additional 3 digits on top of the current approach.
> 
> Note there is also one e2e test failure about region matching. Merging dev to this branch should make the test passing.

Hi Kusong, that is unfortunate, this was working fine before. It is probably due to the recently requested changes I made. I will take a look at them tomorrow, revert some of the changes and hopefully get it working for you again. Is there a way I can see that the e2e tests are failing? That way I can only ping you after they are all passing.

kswang1029 commented:
e2e tests are currently run manually for a fair of frontend+backend commits. They are run automatically on dev+dev. 

Please refer to the two images adopted in the e2e tests for reproducing the issues. (M17 for frequency, HD163296 for velocity)
https://drive.google.com/file/d/1PQRptRzYEWvgN081mIMX-ZytXJ68gV6X/view?usp=sharing
https://drive.google.com/file/d/1GBkof53vxeJb4GfELAj1_uIwJzOyxJqf/view?usp=sharing

jrhosk commented:
> e2e tests are currently run manually for a fair of frontend+backend commits. They are run automatically on dev+dev.
> 
> Please refer to the two images adopted in the e2e tests for reproducing the issues. (M17 for frequency, HD163296 for velocity) https://drive.google.com/file/d/1PQRptRzYEWvgN081mIMX-ZytXJ68gV6X/view?usp=sharing https://drive.google.com/file/d/1GBkof53vxeJb4GfELAj1_uIwJzOyxJqf/view?usp=sharing

Thanks Kusong, I'll work this out this week.

jrhosk committed:
Rolled back changes to the cursor decimal precision to fix issue regarding e2e test failures.

jrhosk commented:
> > e2e tests are currently run manually for a fair of frontend+backend commits. They are run automatically on dev+dev.
> > Please refer to the two images adopted in the e2e tests for reproducing the issues. (M17 for frequency, HD163296 for velocity) https://drive.google.com/file/d/1PQRptRzYEWvgN081mIMX-ZytXJ68gV6X/view?usp=sharing https://drive.google.com/file/d/1GBkof53vxeJb4GfELAj1_uIwJzOyxJqf/view?usp=sharing
> 
> Thanks Kusong, I'll work this out this week.

Alright, hopefully it will pass now. I didn't see any issues using the files you posted.

YuHsuan-Hwang commented on src/utilities/units.ts:
```diff
@@ -96,7 +96,7 @@ export function getVariablePrecision(value: number): number {
 
     let decimalPlacement = 0.1;
-    let precision = 1;
+    let precision = 3;
```
@jrhosk I would prefer we add a comment here, explaining that the precision here is 3 instead of 1 because we add extra 2 digits for better UX.
Could you also update this branch with the latest dev, and add an item to `Added` in CHANGELOG.md?

jrhosk committed:
Add comment to explain additional precision on spectral profile cursor info.

jrhosk committed:
Merge dev into branch.

jrhosk committed:
Forgot CHANGELOG after merge.

jrhosk committed:
Run reformat.

kswang1029 commented:
I spotted a frontend crash.
![Screen Shot 2022-08-05 at 08 30 30](https://user-images.githubusercontent.com/20819712/182977235-de461de4-9a9a-42c0-803f-0fd1c2500a62.png)

To reproduce:
1) load an image
2) make a region and enable a spectral profiler to see the spectral profile
3) without changing anything (not even enable the 2nd spec info), try to zoom the profile all the way in
4) frontend will crash

jrhosk committed:
Fixed issue where an exception is raised when zooming into spectral profiler to the point only one point is shown.

jrhosk commented:
> I spotted a frontend crash. ![Screen Shot 2022-08-05 at 08 30 30](https://user-images.githubusercontent.com/20819712/182977235-de461de4-9a9a-42c0-803f-0fd1c2500a62.png)
> 
> To reproduce:
> 
> 1. load an image
> 2. make a region and enable a spectral profiler to see the spectral profile
> 3. without changing anything (not even enable the 2nd spec info), try to zoom the profile all the way in
> 4. frontend will crash

@kswang1029 I think I fixed this problem. The issue is that the precision is based on the difference between two consecutive tick points, when zooming in to the point where only one point is still shown, there is no "difference" to be calculated. The fix I implemented just checks the data length being displayed and when there is only one point,  it defaults to a precision of 1e-9. A better fix might be to limit the zoom to two points but that might be a bigger fix. Not sure what the use case is for zooming in to only 1 or even 0 points, which you can do. That would be my preferred fix.

kswang1029 commented:
> > I spotted a frontend crash. ![Screen Shot 2022-08-05 at 08 30 30](https://user-images.githubusercontent.com/20819712/182977235-de461de4-9a9a-42c0-803f-0fd1c2500a62.png)
> > To reproduce:
> > 
> > 1. load an image
> > 2. make a region and enable a spectral profiler to see the spectral profile
> > 3. without changing anything (not even enable the 2nd spec info), try to zoom the profile all the way in
> > 4. frontend will crash
> 
> @kswang1029 I think I fixed this problem. The issue is that the precision is based on the difference between two consecutive tick points, when zooming in to the point where only one point is still shown, there is no "difference" to be calculated. The fix I implemented just checks the data length being displayed and when there is only one point, it defaults to a precision of 1e-9. A better fix might be to limit the zoom to two points but that might be a bigger fix. Not sure what the use case is for zooming in to only 1 or even 0 points, which you can do. That would be my preferred fix.

The spectral profiler allows multi-profile plotting. Sometimes we compare profiles with quite different channel widths so it is possible and necessary to zoom profiles at such high zoom levels.

jrhosk commented:
> > > I spotted a frontend crash. ![Screen Shot 2022-08-05 at 08 30 30](https://user-images.githubusercontent.com/20819712/182977235-de461de4-9a9a-42c0-803f-0fd1c2500a62.png)
> > > To reproduce:
> > > 
> > > 1. load an image
> > > 2. make a region and enable a spectral profiler to see the spectral profile
> > > 3. without changing anything (not even enable the 2nd spec info), try to zoom the profile all the way in
> > > 4. frontend will crash
> > 
> > 
> > @kswang1029 I think I fixed this problem. The issue is that the precision is based on the difference between two consecutive tick points, when zooming in to the point where only one point is still shown, there is no "difference" to be calculated. The fix I implemented just checks the data length being displayed and when there is only one point, it defaults to a precision of 1e-9. A better fix might be to limit the zoom to two points but that might be a bigger fix. Not sure what the use case is for zooming in to only 1 or even 0 points, which you can do. That would be my preferred fix.
> 
> The spectral profiler allows multi-profile plotting. Sometimes we compare profiles with quite different channel widths so it is possible and necessary to zoom profiles at such high zoom levels.

That explains it, good to know.

kswang1029 commented:
@jrhosk when there are multiple profiles from matched cubes, we see the 2nd spectral info is Null. In the video I am setting the three matched images as active one by one. 

https://user-images.githubusercontent.com/20819712/183230441-5ab73435-7c23-4f8e-afbc-48bad025c7ab.mov

test images are attached 
https://drive.google.com/file/d/1xrvn0Xss2h_ns-zQh2PPBWD-Bo1Ztdk4/view?usp=sharing

jrhosk commented:
> @jrhosk when there are multiple profiles from matched cubes, we see the 2nd spectral info is Null. In the video I am setting the three matched images as active one by one.
> 
>  Screen.Recording.2022-08-06.at.10.40.53.mov 
> test images are attached https://drive.google.com/file/d/1xrvn0Xss2h_ns-zQh2PPBWD-Bo1Ztdk4/view?usp=sharing

Hi @kswang1029 I can't reproduce this error. I opened all three files you sent (one one and append the other two) then I made a region with the rect selection box and used the wcs matching. I opened the spectral profile window and selected _image_ to get all three profiles showing and added the second coordinate. When zooming in and scrolling through the profile as you did, I don't see and discontinuity.  I don't see a null value anywhere on the profile. Could you post a detailed, step-by-step on how to reproduce this issue please?

kswang1029 commented:
@jrhosk please try the following steps:
1. load HD163296_CO_2_1.fits, HD163296_13CO_2-1.fits and HD163296_C18O_2-1.fits in the sequence
2. match them
3. make a region to see a region spectral profile 
4. enable "Image" multi-profile plot and we will see three profiles
5. zoom to about -16km/s to +2km/s
6. enable the 2nd spectral info and set it to frequency
7. set HD163296_CO_2_1.fits as the active image and check cursor info (this appears fine)
8. set HD163296_13CO_2-1.fits as the active image and we see null (esp. ~0:13 in the video where at the cursor location there are three profiles)
9. set HD163296_C18O_2-1.fits as the active image and we see null


https://user-images.githubusercontent.com/20819712/188762185-e67a2d38-c4a8-4bd7-8462-d731ab71a1c2.mov

veggiesaurus commented on src/components/Shared/SpectralSettings/SpectralSettingsComponent.tsx:
```diff
@@ -39,6 +42,17 @@ export class SpectralSettingsComponent extends React.Component<{
+                            disabled={disableCoordinateSetting}
+                            value={frame && frame.spectralCoordinateSecondary ? frame.spectralCoordinateSecondary : ""}
+                            options={spectralCoordinateOptions}
+                            onChange={(event: React.FormEvent<HTMLSelectElement>) => this.props.onSpectralCoordinateChangeSecondary(event.currentTarget.value as string)}
```
can use implicit typing here (`{event=>this.props.onSpectral...}`)

jrhosk replied:
Would you want all the similar lines to be this way as well? All I did was copy the style of whoever wrote the rest of the code.

YuHsuan-Hwang replied:
applied implicit typing

veggiesaurus commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,22 +153,50 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+            let primaryXStr: string = "";
+
+            // We calculate the difference between neighboring values to get and estimate of the precision.
+            diffLeft = data.length === 1 ? 1e-9 : nearest.index > 0 ? Math.abs(data[nearest.index].x - data[nearest.index - 1].x) : Math.abs(data[nearest.index + 1].x - data[nearest.index].x);
```
I think this would be more readable if you assigned variables to the data points themselves (e.g `const currentData = data[nearest.index]`), but just a suggestion :innocent:

jrhosk replied:
Well yes but what about `data[nearest.index + 1]` and `data[nearest.index - 1]`?  Would you want a new name for every index used?

YuHsuan-Hwang replied:
applied `currentIndex` and `neighborIndex` for better readability

veggiesaurus commented on src/stores/widgets/SpectralProfileWidget/SpectralProfileWidgetStore.ts:
```diff
@@ -34,6 +34,7 @@ export type MultiPlotData = {
     yMax: number;
     yMean: number;
     yRms: number;
+    startEndIndexes: {startIndex: number; endIndex: number}[];
```
this is a very confusing variable name :thinking:

jrhosk replied:
I agree and I am not sure how ti came to be honestly. Maybe a find-replace error on my part. I changed it to `dataIndexes`

veggiesaurus commented on src/utilities/units.ts:
```diff
@@ -43,6 +43,84 @@ export function toFixed(val: number, decimals: number = 0): string {
+}
+
+export function trimTrailingZeros(value: string): string {
+    // Trims the trailing zeros from the input decimal value. If
```
this seem like something that could be done quite easily with a regex replace using something like `/\d?\.\d+?(0+)/gm` (@confluence will have a better solution :innocent: )

YuHsuan-Hwang replied:
applied regex replace in `trimTrailingDecimals` (`trimTrailingZeros` is unused and is removed)

jrhosk committed:
Change the way secondary information is processed.

jrhosk committed:
Changed teh way secondary data is loaded to be sure data from each profile is added instead of basing it on the current frame.

jrhosk committed:
Fixed issue where secondary info was not being updated for all profiles when multiple images were matched.

YuHsuan-Hwang committed:
move data process from component to store

YuHsuan-Hwang committed:
remove unused start/end indexes

YuHsuan-Hwang committed:
adjust genCursoInfoString

YuHsuan-Hwang committed:
use implicit typing

YuHsuan-Hwang committed:
move event handling from store to component

YuHsuan-Hwang committed:
resolve conflicts and merge dev

YuHsuan-Hwang committed:
adjust changelog

YuHsuan-Hwang committed:
rename to secondaryXData

YuHsuan-Hwang committed:
adjust getVariablePrecision; use regex replace for trimTrailingDecimals

YuHsuan-Hwang commented:
The information of this PR is updated in the first comment.

New applied changes:
* moved data processing from component to store
* renamed `secondary`/`secondaryData` to `secondaryXData`
* resolved conflict with the smoothed data cursor info feature. Please note that currently we do not support secondary unit for the smoothed data. This needs extra calculation, and I prefer we file a new issue if needed.
* adjustments according to the comments: applied implicit typing in `SpectralSettingsComponent`; `currentIndex` and `neighborIndex` in `SpectralProfilerComponent` for better readability; regex replace in `trimTrailingDecimals`

YuHsuan-Hwang commented:
@kswang1029 This is ready for testing, thanks.

YuHsuan-Hwang commented:
fixing formatting issue observed by @kswang1029

acdo2002 reviewed (approved):
I have tested that opening the four images and set four different combinations of first/second coordinates. So far the new function (the first and the second coordinates) works fine and smooth.

YuHsuan-Hwang committed:
fix trimTrailingDecimals

YuHsuan-Hwang committed:
adjust column width

YuHsuan-Hwang committed:
fix incorrect slicing

YuHsuan-Hwang committed:
fix conversion with native unit

YuHsuan-Hwang commented:
new changes:
* fixed incorrect regex replace for removing trailing zeros
* fixed incorrect slicing; now null will not be shown at the end of data
* fixed conversion when primary unit is native unit
* adjusted the width of column in the setting panel

kswang1029 reviewed (approved):
@YuHsuan-Hwang this now works nicely! 👍  No regression is observed via current e2e tests.

YuHsuan-Hwang committed:
Merge branch 'dev' into jrhosk/1837-optional-spectral-profiler-info

crocka commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,25 +152,40 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+        return spectralType === SpectralType.CHANNEL ? toFixed(data) : toFormattedNotationByDiff(data, diff);
+    };
+
+    private genCursoInfoString = (data: Point2D[], smoothedData: Point2D[], secondaryXData: number[], cursorXValue: number, cursorXUnit: string, label: string): string => {
```
Is this a typo? Is it genCursorInfoString?

kswang1029 replied:
Eagle eye 🦅

YuHsuan-Hwang replied:
good catch! Typo fixed.

crocka commented on src/components/SpectralProfiler/SpectralProfilerComponent.tsx:
```diff
@@ -152,25 +152,40 @@ export class SpectralProfilerComponent extends React.Component<WidgetProps> {
+                const secondaryXStr = this.precisionFormatting(secondaryXData[currentIndex], diffLeft, frame.spectralTypeSecondary);
+
+                if (frame.spectralTypeSecondary !== SpectralType.CHANNEL) secondaryXUnit = frame.spectralUnitSecondary;
+                else secondaryChannelString = "Channel";
```
Angus has told me in my previous PR that he wants to always have brackets for if-statements.

YuHsuan-Hwang replied:
fixed

YuHsuan-Hwang committed:
fix typo; add brackets
